# NWM Assignment
Name: Arvind Ganesh

In [761]:
import numpy as np
import pandas as pd

• Main file containing election data

In [762]:
df = pd.read_csv('Bihar_Election_Results.csv', index_col=None)

In [763]:
print(df)

      Election Year  Position                     Name  Votes Votes %  \
0              2010         1             Rajesh Singh  42289  29.40%   
1              2010         2    Mukesh Kumar Kushwaha  27618  19.20%   
2              2010         3  Dheerendra Pratap Singh  20886  14.50%   
3              2010         4           Irshad Hussain  17747  12.40%   
4              2010         5      Deep Narayan Mahato  14047   9.80%   
...             ...       ...                      ...    ...     ...   
12296      2005-Feb        15  Chandeshwar Prasad Gupt    534   0.60%   
12297      2005-Feb        16             Tirpit Singh    498   0.60%   
12298      2005-Feb        17        Surendra Rajbansi    437   0.50%   
12299      2005-Feb        18       Nawin Kumar Paswan    415   0.50%   
12300      2005-Feb        19               Ajay Kumar    407   0.50%   

                          Party        AC name  AC No  
0           Janata Dal (United)  Valmiki Nagar      1  
1          

In [764]:
df.shape

(12301, 8)

In [765]:
df.columns

Index(['Election Year', 'Position', 'Name', 'Votes', 'Votes %', 'Party',
       'AC name', 'AC No'],
      dtype='object')

In [766]:
df['Election Year'].unique()

array(['2010', '2015', '2005-Oct', '2005-Feb'], dtype=object)

• There are 4 different elections involved.

In [767]:
b=np.sort(df['Party'].unique())
for i in range(len(b)):
    print(b[i])

Aadarsh Mithila Party
Aam Janata Party
Aap Aur Hum Party
Aarakshan Virodhi Party
Adarsh Political Party
Akhand Bharat Samaj Party
Akhand Jharkhand People's Front
Akhil Bharat Hindu Mahasabha
Akhil Bharatiya Ashok Sena
Akhil Bharatiya Desh Bhakt Morcha
Akhil Bharatiya Hind Kranti Party
Akhil Bharatiya Jan Sangh
Akhil Bharatiya Rajarya Sabha
Akhil Bharitya Mithila Party
Akhil Bhartiya Atyant Pichara Sangharsh Morcha Party
Akhil Bhartiya Mithila Party
Akhil Hind Forward Block
Akhil Jan Vikas Dal
All India Babu Jagjivan Ram Saheb National Congress
All India Forward Bloc
All India Forward Bloc (Subhasist)
All India Majlis-E-Ittehadul Muslimoon
Alpjan Samaj Party
Ambedkar National Congress
Amra Bangalee
Apna Dal
Atulya Bharat Party
Awami Party
Bahujan Ekta Party (R)
Bahujan Kranti Party (Marxwad-Ambedkarwad)
Bahujan Mukti Party
Bahujan Nyay Dal
Bahujan Samaj Party
Bahujan Sangharsh Party (Kanshiram)
Bahujan Shakty
Bajjikanchal Vikas Party
Bharat Bhrashtachar Mitao Party
Bharat Deshbhakt Part

In [768]:
np.shape(df['Party'].unique())

(222,)

In [769]:
df['AC name'].unique()

array(['Valmiki Nagar', 'Ramnagar', 'Narkatiaganj', 'Bagaha', 'Lauriya',
       'Nautan', 'Chanpatia', 'Bettiah', 'Sikta', 'Raxaul', 'Sugauli',
       'Narkatia', 'Harsidhi', 'Govindganj', 'Kesaria', 'Kalyanpur',
       'Pipra', 'Madhuban', 'Motihari', 'Chiraia', 'Dhaka', 'Sheohar',
       'Riga', 'Bathnaha', 'Parihar', 'Sursand', 'Bajpatti', 'Sitamarhi',
       'Runisaidpur', 'Belsand', 'Harlakhi', 'Benipatti', 'Khajauli',
       'Babubarhi', 'Bisfi', 'Madhubani', 'Rajnagar', 'Jhanjharpur',
       'Phulparas', 'Laukaha', 'Nirmali', 'Supaul', 'Tribeniganj',
       'Chhatapur', 'Narpatganj', 'Raniganj', 'Forbesganj', 'Araria',
       'Jokihat', 'Sikti', 'Bahadurganj', 'Thakurganj', 'Kishanganj',
       'Kochadhaman', 'Amour', 'Baisi', 'Kasba', 'Banmankhi', 'Rupauli',
       'Dhamdaha', 'Purnia', 'Katihar', 'Kadwa', 'Balrampur', 'Pranpur',
       'Manihari', 'Barari', 'Korha', 'Alamnagar', 'Bihariganj',
       'Singheshwar', 'Madhepura', 'Sonbarsa', 'Saharsa',
       'Simri Bakhtiarpur',

In [770]:
np.shape(df['AC name'].unique())

(289,)

• From the above, we expect there are 289 constituencies in the data.

In [771]:
df['AC No'].unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,  53,
        54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,
        67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,
        80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,
        93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104, 105,
       106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118,
       119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 132,
       133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145,
       146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158,
       159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171,
       172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 18

• But there are only 243 indices corresponding to the constituencies.
• We will see why.

In [772]:
np.shape(df['AC No'].unique())

(243,)

In [773]:
df.loc[df['AC No'] == 2, 'AC name'].unique()

array(['Ramnagar', 'Ramnagar (SC)'], dtype=object)

• There are two reasons.
• First, SC and ST data are separated from others. THIS we keep.
• Second, different spelling for AC name are used throughout.
• For example, 'Bibhutpur' and 'Bibhutipur'. THIS we eliminate.

In [774]:
for i in range(1, np.shape(df['AC No'].unique())[0] + 1):
    a = df.loc[df['AC No'] == i, 'AC name'].unique()
    print(a)

['Valmiki Nagar']
['Ramnagar' 'Ramnagar (SC)']
['Narkatiaganj']
['Bagaha']
['Lauriya']
['Nautan']
['Chanpatia']
['Bettiah']
['Sikta']
['Raxaul']
['Sugauli']
['Narkatia']
['Harsidhi' 'Harsidhi (SC)']
['Govindganj']
['Kesaria']
['Kalyanpur' 'Kalyanpur (SC)']
['Pipra']
['Madhuban']
['Motihari']
['Chiraia']
['Dhaka']
['Sheohar']
['Riga']
['Bathnaha' 'Bathnaha (SC)']
['Parihar']
['Sursand']
['Bajpatti']
['Sitamarhi']
['Runisaidpur' 'Runnisaidpur']
['Belsand']
['Harlakhi']
['Benipatti']
['Khajauli']
['Babubarhi']
['Bisfi']
['Madhubani']
['Rajnagar' 'Rajnagar (SC)']
['Jhanjharpur']
['Phulparas']
['Laukaha']
['Nirmali']
['Pipra']
['Supaul']
['Tribeniganj' 'Triveniganj (SC)']
['Chhatapur']
['Narpatganj']
['Raniganj' 'Raniganj (SC)']
['Forbesganj']
['Araria']
['Jokihat']
['Sikti']
['Bahadurganj']
['Thakurganj']
['Kishanganj']
['Kochadhaman']
['Amour']
['Baisi']
['Kasba']
['Banmankhi' 'Banmankhi (SC)']
['Rupauli']
['Dhamdaha']
['Purnia']
['Katihar']
['Kadwa']
['Balrampur']
['Pranpur']
['Manihari'

# DATA PREPROCESSING
1) Correcting 'AC name' spellings

In [775]:
for i in range(1, np.shape(df['AC No'].unique())[0] + 1):
    AC = df.loc[df['AC No'] == i, 'AC name'].unique()
    if AC.size == 2 and 'SC' not in AC[1] and 'ST' not in AC[1]:
        df['AC name'] = df['AC name'].replace([ AC[1] ], AC[0])

In [776]:
df['AC name'].unique().size

281

• Size reduced from 289 to 281

In [777]:
for i in range(1, np.shape(df['AC No'].unique())[0] + 1):
    a = df.loc[df['AC No'] == i, 'AC name'].unique()
    print(a)

['Valmiki Nagar']
['Ramnagar' 'Ramnagar (SC)']
['Narkatiaganj']
['Bagaha']
['Lauriya']
['Nautan']
['Chanpatia']
['Bettiah']
['Sikta']
['Raxaul']
['Sugauli']
['Narkatia']
['Harsidhi' 'Harsidhi (SC)']
['Govindganj']
['Kesaria']
['Kalyanpur' 'Kalyanpur (SC)']
['Pipra']
['Madhuban']
['Motihari']
['Chiraia']
['Dhaka']
['Sheohar']
['Riga']
['Bathnaha' 'Bathnaha (SC)']
['Parihar']
['Sursand']
['Bajpatti']
['Sitamarhi']
['Runisaidpur']
['Belsand']
['Harlakhi']
['Benipatti']
['Khajauli']
['Babubarhi']
['Bisfi']
['Madhubani']
['Rajnagar' 'Rajnagar (SC)']
['Jhanjharpur']
['Phulparas']
['Laukaha']
['Nirmali']
['Pipra']
['Supaul']
['Tribeniganj' 'Triveniganj (SC)']
['Chhatapur']
['Narpatganj']
['Raniganj' 'Raniganj (SC)']
['Forbesganj']
['Araria']
['Jokihat']
['Sikti']
['Bahadurganj']
['Thakurganj']
['Kishanganj']
['Kochadhaman']
['Amour']
['Baisi']
['Kasba']
['Banmankhi' 'Banmankhi (SC)']
['Rupauli']
['Dhamdaha']
['Purnia']
['Katihar']
['Kadwa']
['Balrampur']
['Pranpur']
['Manihari' 'Manihari (ST)

2) All party names are made lowercase because there is discrepancy in data.

In [778]:
df['Party'] = df['Party'].str.lower()

In [779]:
df['Party']

0             janata dal (united)
1            rashtriya janata dal
2             bahujan samaj party
3        indian national congress
4                     independent
                   ...           
12296                 independent
12297                 independent
12298                 independent
12299                 independent
12300                 independent
Name: Party, Length: 12301, dtype: object

3) Further corrections based on observed discrepancies.

In [784]:
df['Party'] = df['Party'].replace(['akhil bharitya mithila party'],'akhil bhartiya mithila party')
df['Party'] = df['Party'].replace(['proutist bloc india'],'proutist bloc of india')
df['Party'] = df['Party'].replace(['proutist sarva samaj party'],'proutist sarva samaj')
df['Party'] = df['Party'].replace(['janvadi party(socialist)'],'janvadi party (socialist)')

In [785]:
df['Party'].unique().size

217

In [786]:
parties=np.sort(df['Party'].unique())
for i in range(len(parties)):
    print(parties[i])

aadarsh mithila party
aam janata party
aap aur hum party
aarakshan virodhi party
adarsh political party
akhand bharat samaj party
akhand jharkhand people's front
akhil bharat hindu mahasabha
akhil bharatiya ashok sena
akhil bharatiya desh bhakt morcha
akhil bharatiya hind kranti party
akhil bharatiya jan sangh
akhil bharatiya rajarya sabha
akhil bhartiya atyant pichara sangharsh morcha party
akhil bhartiya mithila party
akhil hind forward block
akhil jan vikas dal
all india babu jagjivan ram saheb national congress
all india forward bloc
all india forward bloc (subhasist)
all india majlis-e-ittehadul muslimoon
alpjan samaj party
ambedkar national congress
amra bangalee
apna dal
atulya bharat party
awami party
bahujan ekta party (r)
bahujan kranti party (marxwad-ambedkarwad)
bahujan mukti party
bahujan nyay dal
bahujan samaj party
bahujan sangharsh party (kanshiram)
bahujan shakty
bajjikanchal vikas party
bharat bhrashtachar mitao party
bharat deshbhakt party
bharat ka kisan majdoor par

• total_votes tells you total votes for each party in each election.

In [787]:
yunique = df['Election Year'].unique()
punique = df['Party'].unique()
len1 = len(yunique)
len2 = len(punique)
total_votes = np.zeros((len1, len2))
for index1 in range(len1):
    #print(index1)
    #print(df['Election Year'][index1])
    for index2 in range(len2):
        total_votes[index1, index2] = df.loc[(df['Election Year'] == yunique[index1]) & (df['Party'] == punique[index2]), 'Votes'].sum()

In [788]:
yunique

array(['2010', '2015', '2005-Oct', '2005-Feb'], dtype=object)

• total_votes_seats tells you total votes for each party in each election in each constituency.

In [792]:
cunique = df['AC name'].unique()
len3 = len(cunique)
total_votes_seats = np.zeros((len1, len2, len3))
for index1 in range(len1):
    for index2 in range(len2):
        #print(index2)
        for index3 in range(len3):
            total_votes_seats[index1, index2, index3] = df.loc[(df['Election Year'] == yunique[index1]) & (df['Party'] == punique[index2]) & (df['AC name'] == cunique[index3]), 'Votes'].sum()

• seat_winner tells which party won each constituency in each election.

In [793]:
seat_winner = pd.DataFrame(np.argmax(total_votes_seats, axis=1))

In [794]:
seat_winner.shape

(4, 281)

• unique_winners_AC tells you no of distinct parties which held seat at each constituency

In [795]:
unique_winners_AC = seat_winner.nunique()
unique_winners_AC

0      3
1      2
2      2
3      3
4      3
      ..
276    2
277    2
278    2
279    2
280    2
Length: 281, dtype: int64

# 1a. Election winners

In [789]:
overall_election_winner

array([0, 8, 1, 1])

In [790]:
overall_election_winner = np.argmax(total_votes, axis=1)
#punique[overall_election_winner]

dictionary = dict(zip(yunique, punique[overall_election_winner]))
dictionary

{'2010': 'janata dal (united)',
 '2015': 'bharatiya janta party',
 '2005-Oct': 'rashtriya janata dal',
 '2005-Feb': 'rashtriya janata dal'}

In [ ]:
unique_winners_AC = seat_winner.nunique()
unique_winners_AC

# 1b. Swing seats

In [796]:
swing = np.zeros(seat_winner.shape[1])
for i in range(len(swing)):
    swing[i] = (unique_winners_AC[i] == seat_winner.shape[0])

In [797]:
cunique[np.where(swing==1)]

array(['Riga', 'Belsand', 'Madhubani', 'Thakurganj', 'Amour', 'Kadwa',
       'Balrampur', 'Darauli', 'Lalganj', 'Raghopur', 'Teghra',
       'Begusarai', 'Barbigha', 'Ramgarh', 'Sherghati', 'Jhajha',
       'Chakai'], dtype=object)

# 1c. Top 8 parties in each election

In [798]:
total_votes.shape

(4, 217)

In [807]:
top_parties = np.argsort(-total_votes,axis=1)
top_8_parties = np.transpose(top_parties)[:8,:]
pd.DataFrame(punique[top_8_parties],columns=[yunique])

2010                              2015  \
0         janata dal (united)             bharatiya janta party   
1        rashtriya janata dal              rashtriya janata dal   
2       bharatiya janta party               janata dal (united)   
3                 independent                       independent   
4    indian national congress          indian national congress   
5        lok jan shakti party              lok jan shakti party   
6         bahujan samaj party         rashtriya lok samta party   
7  nationalist congress party  hindustani awam morcha (secular)   

                   2005-Oct                  2005-Feb  
0      rashtriya janata dal      rashtriya janata dal  
1       janata dal (united)               independent  
2     bharatiya janta party       janata dal (united)  
3      lok jan shakti party      lok jan shakti party  
4               independent     bharatiya janta party  
5  indian national congress  indian national congress  
6       bahujan samaj party       bahujan samaj party  
7           samajwadi party           samajwadi party

# 1d. Minimum votes of top 8 from 2015, in past for each constituency

In [799]:
party_index = np.array([0] + list(range(2,len1)))
party_index

array([0, 2, 3])

In [817]:
party_index = np.array([0] + list(range(2,len1)))
min_2015_AC = np.amin(total_votes_seats[np.ix_(party_index,np.transpose(top_8_parties)[1,:].astype(int))],axis=0)
min_2015_AC.shape
# convert to dataframe
min_2015_AC = pd.DataFrame(np.transpose(min_2015_AC).astype(int),columns = [punique[top_8_parties][:,1]],index = df['AC name'].unique())

In [818]:
min_2015_AC

bharatiya janta party rashtriya janata dal  \
Valmiki Nagar                       0                19164   
Ramnagar                            0                    0   
Narkatiaganj                    20961                 8243   
Bagaha                              0                    0   
Lauriya                             0                    0   
...                               ...                  ...   
Imamganj (SC)                       0                    0   
Barachatti (SC)                     0                    0   
Bodh Gaya (SC)                      0                    0   
Rajauli (SC)                        0                    0   
Sikandra (SC)                       0                    0   

                janata dal (united) independent indian national congress  \
Valmiki Nagar                  1716        2829                        0   
Ramnagar                          0           0                        0   
Narkatiaganj                      0        6050                        0   
Bagaha                            0        6147                        0   
Lauriya                           0        5154                     1757   
...                             ...         ...                      ...   
Imamganj (SC)                     0           0                        0   
Barachatti (SC)                   0           0                        0   
Bodh Gaya (SC)                    0           0                        0   
Rajauli (SC)                      0           0                        0   
Sikandra (SC)                     0           0                        0   

                lok jan shakti party rashtriya lok samta party  \
Valmiki Nagar                      0                         0   
Ramnagar                           0                         0   
Narkatiaganj                       0                         0   
Bagaha                             0                         0   
Lauriya                            0                         0   
...                              ...                       ...   
Imamganj (SC)                      0                         0   
Barachatti (SC)                    0                         0   
Bodh Gaya (SC)                     0                         0   
Rajauli (SC)                       0                         0   
Sikandra (SC)                      0                         0   

                hindustani awam morcha (secular)  
Valmiki Nagar                                  0  
Ramnagar                                       0  
Narkatiaganj                                   0  
Bagaha                                         0  
Lauriya                                        0  
...                                          ...  
Imamganj (SC)                                  0  
Barachatti (SC)                                0  
Bodh Gaya (SC)                                 0  
Rajauli (SC)                                   0  
Sikandra (SC)                                  0  

[281 rows x 8 columns]

# 1e. Tripolar constituencies

In [819]:
tvs_sum = total_votes_seats.sum(axis=1)
tvs_sum[tvs_sum == 0] = 1
#np.max(tvs_sum)

In [820]:
norm_tvs = total_votes_seats / tvs_sum[:,np.newaxis,:]
print(total_votes_seats.shape)
norm_tvs.shape

(4, 217, 281)


(4, 217, 281)

In [824]:
norm_tvs = -np.sort(-norm_tvs,axis=1)
#norm_tvs

In [835]:
tripolar=-1*np.ones(len3)
tripolar = np.all(np.all(norm_tvs[:,0:3,:] > 0.1, axis = 1) == True, axis = 0)
#np.where(tripolar==True)

In [834]:
pd.DataFrame(cunique[tripolar])

0
0       Valmiki Nagar
1        Narkatiaganj
2             Lauriya
3             Sugauli
4            Narkatia
5           Kalyanpur
6               Pipra
7             Sheohar
8            Harlakhi
9         Bahadurganj
10        Kochadhaman
11              Baisi
12           Dhamdaha
13            Katihar
14              Kadwa
15          Balrampur
16            Pranpur
17             Barari
18          Alamnagar
19  Simri Bakhtiarpur
20        Gora Bauram
21            Gaighat
22              Kanti
23              Siwan
24            Ziradei
25             Chapra
26             Amnour
27            Lalganj
28         Warisnagar
29           Ujiarpur
30              Morwa
31          Bibhutpur
32             Teghra
33          Kahalgaon
34         Sultanganj
35              Banka
36         Sheikhpura
37             Tarari
38         Jagdishpur
39            Ramgarh
40             Bhabua
41           Chainpur
42           Kargahar
43              Dehri
44            Karakat
45              Arwal
46          Nabinagar
47         Aurangabad

# 1f. Stronghold seats

In [840]:
stronghold_index = np.zeros(len3)
for index in range(len3):
    stronghold_index[index] = (unique_winners_AC[index] == 1)
    
#stronghold_index

In [841]:
pd.DataFrame(cunique[np.where(unique_winners_AC == 1)])

0
0          Chanpatia
1            Bettiah
2            Parihar
3            Nirmali
4        Singheshwar
5              Siwan
6        Maharajganj
7         Warisnagar
8           Khagaria
9           Gopalpur
10          Dhuraiya
11           Nalanda
12            Rajpur
13            Kurtha
14     Ramnagar (SC)
15     Harsidhi (SC)
16  Singheshwar (SC)
17       Garkha (SC)
18    Dhauraiya (SC)
19       Rajpur (SC)
20      Mohania (SC)

# 1g. Close contest constituencies

In [842]:
close_contest = np.all(np.abs(np.diff(norm_tvs[:,0:2,:], axis=1)) < 0.1, axis=0)

In [851]:
pd.DataFrame(cunique[np.where(close_contest==True)[1]])

0
0                   Raxaul
1                  Parihar
2                Sitamarhi
3                 Harlakhi
4                Madhubani
5                 Rajnagar
6               Thakurganj
7                    Kasba
8                  Pranpur
9                    Keoti
10                 Kurhani
11                 Ziradei
12                 Darauli
13            Raghunathpur
14               Pirpainti
15               Bhagalpur
16              Sultanganj
17               Nathnagar
18                   Banka
19                  Belhar
20              Jagdishpur
21                 Ramgarh
22                 Mohania
23                Kargahar
24                 Karakat
25                     Goh
26                  Nawada
27           Bathnaha (SC)
28              Korha (SC)
29        Singheshwar (SC)
30  Kusheshwar Asthan (SC)
31            Bochaha (SC)
32              Sakra (SC)
33             Bhorey (SC)
34         Raja Pakar (SC)
35            Patepur (SC)
36             Rosera (SC)
37             Bakhri (SC)
38             Alauli (SC)
39          Pirpainti (SC)
40          Dhauraiya (SC)
41             Rajgir (SC)
42             Agiaon (SC)
43             Rajpur (SC)
44            Mohania (SC)
45            Chenari (SC)
46       Makhadumapur (SC)
47            Kutumba (SC)
48           Sikandra (SC)

# 1h. Growing seats

In [878]:
idx=np.array([2,3,1,0])
party_gs = np.where(np.all(np.diff(total_votes_seats[np.ix_(idx,top_8_parties[:,3])],axis=0) > 0, axis=0)==True)
party_gs

growing_seats = [[],[],[],[],[],[],[],[]]
for index in range(len(party_gs[0])):
    growing_seats[party_gs[0][index]].append(cunique[party_gs[1][index]])

(array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 2, 4, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6]),
 array([ 18,  28,  29,  46,  48,  63,  73,  75,  83,  86,  91,  95, 104,
        107, 109, 113, 115, 124, 127, 147, 148, 163, 165, 167, 168, 174,
        183, 205, 207, 223, 216,  54,   3,  11,  98, 115, 118, 160, 174,
        187, 195, 205, 223, 235]))

In [897]:
pd.DataFrame(growing_seats,index=[punique[top_8_parties][:,3]])

0            1        2           3        4   \
rashtriya janata dal          None         None     None        None     None   
independent               Motihari  Runisaidpur  Belsand  Forbesganj  Jokihat   
janata dal (united)            Goh         None     None        None     None   
lok jan shakti party          None         None     None        None     None   
bharatiya janta party        Amour         None     None        None     None   
indian national congress      None         None     None        None     None   
bahujan samaj party         Bagaha     Narkatia  Barauli    Marhaura   Amnour   
samajwadi party               None         None     None        None     None   

                                 5        6         7           8        9   \
rashtriya janata dal           None     None      None        None     None   
independent               Balrampur  Saharsa   Mahishi  Bahadurpur  Gaighat   
janata dal (united)            None     None      None        None     None   
lok jan shakti party           None     None      None        None     None   
bharatiya janta party          None     None      None        None     None   
indian national congress       None     None      None        None     None   
bahujan samaj party          Belhar  Harnaut  Paliganj     Shahpur  Sasaram   
samajwadi party                None     None      None        None     None   

                          ...        20        21          22        23  \
rashtriya janata dal      ...      None      None        None      None   
independent               ...  Parbatta  Jamalpur  Lakhisarai  Barbigha   
janata dal (united)       ...      None      None        None      None   
lok jan shakti party      ...      None      None        None      None   
bharatiya janta party     ...      None      None        None      None   
indian national congress  ...      None      None        None      None   
bahujan samaj party       ...      None      None        None      None   
samajwadi party           ...      None      None        None      None   

                                24       25       26       27      28  \
rashtriya janata dal          None     None     None     None    None   
independent               Asthawan  Harnaut  Danapur  Sasaram  Dinara   
janata dal (united)           None     None     None     None    None   
lok jan shakti party          None     None     None     None    None   
bharatiya janta party         None     None     None     None    None   
indian national congress      None     None     None     None    None   
bahujan samaj party           None     None     None     None    None   
samajwadi party               None     None     None     None    None   

                                 29  
rashtriya janata dal           None  
independent               Sherghati  
janata dal (united)            None  
lok jan shakti party           None  
bharatiya janta party          None  
indian national congress       None  
bahujan samaj party            None  
samajwadi party                None  

[8 rows x 30 columns]

# 1i. Declining seats

In [903]:
idx=np.array([2,3,1,0])
party_ds = np.where(np.all(np.diff(total_votes_seats[np.ix_(idx,top_8_parties[:,3])],axis=0) < 0, axis=0)==True)
print(party_ds)

declining_seats = [[],[],[],[],[],[],[],[]]
for index in range(len(party_ds[0])):
    declining_seats[party_ds[0][index]].append(cunique[party_ds[1][index]])
    

(array([4, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7, 7, 7, 7]), array([211,   8,  31,  34,  68,  77,  86,  92, 130, 137, 168, 206, 209,
       212,   0,   5,   6,  80, 114, 131, 140, 218]))


In [904]:
pd.DataFrame(declining_seats,index=[punique[top_8_parties][:,3]])    

0          1          2   \
rashtriya janata dal               None       None       None   
independent                        None       None       None   
janata dal (united)                None       None       None   
lok jan shakti party               None       None       None   
bharatiya janta party             Arwal       None       None   
indian national congress           None       None       None   
bahujan samaj party               Sikta  Benipatti      Bisfi   
samajwadi party           Valmiki Nagar     Nautan  Chanpatia   

                                       3            4         5            6   \
rashtriya janata dal                 None         None      None         None   
independent                          None         None      None         None   
janata dal (united)                  None         None      None         None   
lok jan shakti party                 None         None      None         None   
bharatiya janta party                None         None      None         None   
indian national congress             None         None      None         None   
bahujan samaj party             Alamnagar  Gora Bauram   Gaighat  Muzaffarpur   
samajwadi party           Darbhanga Rural      Taraiya  Ujiarpur       Teghra   

                                  7         8         9         10     11  \
rashtriya janata dal            None      None      None      None   None   
independent                     None      None      None      None   None   
janata dal (united)             None      None      None      None   None   
lok jan shakti party            None      None      None      None   None   
bharatiya janta party           None      None      None      None   None   
indian national congress        None      None      None      None   None   
bahujan samaj party       Samastipur  Hasanpur  Asthawan  Kargahar  Dehri   
samajwadi party            Nabinagar      None      None      None   None   

                              12  
rashtriya janata dal        None  
independent                 None  
janata dal (united)         None  
lok jan shakti party        None  
bharatiya janta party       None  
indian national congress    None  
bahujan samaj party       Kurtha  
samajwadi party             None

# 1j. State-winner-match constituencies

In [910]:
swm_proxy = (seat_winner.to_numpy() == overall_election_winner[:,None])
swm = np.where(np.all(swm_proxy == True, axis = 0) == True)

cunique[swm]

array(['Bihariganj'], dtype=object)

# x-----------------------x